### DISCUSSION TITLE
Stern Smart Magnet Board integration with MPF and Multimorphic Hardware

### DISCUSSION THREAD

**Discussion by driskel** [missionpinball/missionpinball#76] [Q&A] [answered]
I would like to incorporate the Stern Smart Magnet Processing Board with MPF and Multimorphic hardware P3-Roc, PD-LED, PD-16 and SW-16.   My general question is can I use the PD-LED as the digital outputs for this tutorial:

[https://docs.missionpinball.org/en/stable/mechs/magnets/stern_magnet_pcb.html](url)

**Comment by colemanomartin** [comment]
According to the tutorial the magnet processing board expects 5V for signaling, so 3.3 might not be adequate.  If I understand this optical isolator board correctly it should do level shifting and provide optical isolation.  I would be tempted to use it even if level shifting is not required as the PD-LED is a pricy board https://www.sparkfun.com/products/9118

**Comment by colemanomartin** [comment]
Ay body using the magnet processing board in a homebrew? I'm wondering about the switch output.  Does it pull high or low when a ball grabs? I'm hoping that if the board is appropriately strobed I can connect the switch output to a Multimorphic SW-16 board and read it's state.

**Comment by colemanomartin** [comment, marked as answer]
I got it working.  Here are some notes:
I created a pull request to update the docs with correct info.  
Output is Pin 4.  For output to provide current at 5V sufficient to drive a opto isolator, give pin 3 +5V. To attach this board to a SW-16
use an opto isolator. Spark fun has a good one on a break out board that accepts +5 input and will connect to the SW-16 12V logic.
Pins 5 is "data 6" pin 6 is "data 7" pin 8 is strobe. 
Magnet does not detect ball when ball is on the core.  It needs 3-4 mm of wood between the core and the ball.  It actually detects better with more distance but then the magnet is too weak to hold the ball.
The magnet processing board expects +5V logic.  Depending on how you are driving it, you may need logic level converters.  I controlled it with the PD-LED configuring 3 lights as digital outputs. I then used logic level shifters to bring the 3.3 V from the PD-LED up to 5V to control the board. It might work directly but I did not want to risk cooking my PD-LED. You can't use the PD-16 drivers to control the boards inputs unless you get fancy since the PD-16 does not hand out +5V, it sinks to GND. (Oh, digital outputs on PD-16 flags an error in MPF, probably something syntax related in the programming.  This appears to be a genuine bug.)
I made MPF shows that bitbang the inputs, 200 ms for data 6 and 7 and 100ms for strobe.  I set data lines 6 and 7 for 50 ms before hitting it with a strobe pulse. In other words the strobe pulse is timed to sit inside the data 6 and 7 programming pulses. This is a little different timing than was used in the example. There is need to run the show as a loop (set loops: 0) since the board will stay in the mode until instructed to do something else.  This is a problem if MPF crashes as the board will remain active. Probably it would be better to control the board with a micro controller that would cover this corner case.
